crates/iceberg/src/spec/schema.rs

```diff
@@ -642,6 +644,199 @@ impl SchemaVisitor for IndexByName {
+                return Ok(list.clone());
+            }
+            Ok(ListType {
+                element_field: Arc::new(NestedField {
+                    id: list.element_field.id,
+                    name: list.element_field.name.clone(),
+                    required: list.element_field.required,
+                    field_type: Box::new(value_type),
+                    doc: list.element_field.doc.clone(),
+                    initial_default: list.element_field.initial_default.clone(),
+                    write_default: list.element_field.write_default.clone(),
+                }),
+            })
+        } else {
+            Err(Error::new(ErrorKind::DataInvalid, "Type value is none"))
+        }
+    }
+    fn project_map(map: &MapType, value: Option<Type>) -> Result<MapType> {
+        if let Some(value_type) = value {
+            if *map.value_field.field_type == value_type {
+                return Ok(map.clone());
+            }
+            Ok(MapType {
+                key_field: map.key_field.clone(),
+                value_field: Arc::new(NestedField {
+                    id: map.value_field.id,
+                    name: map.value_field.name.clone(),
+                    required: map.value_field.required,
+                    field_type: Box::new(value_type),
+                    doc: map.value_field.doc.clone(),
+                    initial_default: map.value_field.initial_default.clone(),
+                    write_default: map.value_field.write_default.clone(),
+                }),
+            })
+        } else {
+            Err(Error::new(ErrorKind::DataInvalid, "Type value is none"))
+        }
+    }
+}
+
+impl SchemaVisitor for PruneColumn {
+    type T = Option<Type>;
+
+    fn schema(&mut self, _schema: &Schema, value: Self::T) -> Result<Self::T> {
+        Ok(Some(value.unwrap()))
+    }
+
+    fn field(&mut self, field: &NestedFieldRef, value: Self::T) -> Result<Self::T> {
+        if self.selected.contains(&field.id) {
+            if self.select_full_types {
+                Ok(Some(*field.field_type.clone()))
+            } else if field.field_type.is_struct() {
+                return Ok(Some(Type::Struct(PruneColumn::project_selected_struct(
+                    value,
+                )?)));
+            } else if !field.field_type.is_nested() {
+                return Ok(Some(*field.field_type.clone()));
+            } else {
+                return Err(Error::new(
+                    ErrorKind::DataInvalid,
+                    "Nested field types not allowed in this context".to_string(),
+                ));
+            }
+        } else {
+            Ok(value)
+        }
+    }
+
+    fn r#struct(&mut self, r#struct: &StructType, results: Vec<Self::T>) -> Result<Self::T> {
+        let fields = r#struct.fields();
+        let mut selected_field = Vec::with_capacity(fields.len());
+        let mut same_type = true;
+
+        for i in 0..results.len() {
```

Comment:
We can replace with `https://docs.rs/itertools/latest/itertools/fn.zip_eq.html` to make code easier to read.